### ISSUE TITLE
[HPC] Understand Horovod tutorials

### ISSUE THREAD

**Issue by Setol21** [pescap/FastDeepXDE#4]
Tasks:   

- [x] Reproduce [this section](https://github.com/pescap/EasyHPC/blob/main/docs/source/horovod.rst)
- [ ] Understand `tensorflow2_synthetic_benchmark.py`
- [ ] Understand `tensorflow2_keras_mnist.py` and its code. 
- [ ] Create a benchmark for `tensorflow2_synthetic_benchmark.py` with respect to the number of GPUs. 
- [ ] Add a table with the results

**Comment by Setol21** [comment]
For the `tensorflow2_keras_mnist.py` file:

## General
1. I managed to understand that this code runs in every gpu, therefore when `hvd.rank()` is used it gives the number of the worker (gpu). So when `hvd.rank()==0` appears on the code it means that part of the code belongs only for the worker 0.

## Callbacks:
1. In the part of callbacks, `hvd.callbacks.BroadcastGlobalVariablesCallback(0)` makes sure the initial variables are broadcasted to the rest of workers. It is super important when working with random or checkpoint (previous trained/saved) data. The other command `hvd.callbacks.MetricAverageCallback()` averages the metrics from all workers (to work like a synchronized team).

2. `hvd.callbacks.LearningRateWarmupCallback(initial_lr=scaled_lr, warmup_epochs=3, verbose=1)` This part was an *exclusive* for this code, when training it, it showed worse accuracy when not scaling the learning rate in the first 3 epochs, but I think this is something to come with when searching for a good accuracy.

**Comment by Setol21** [comment]
@pescap 

Benchmarking tensorflow2_keras_synthetic_benchmark.py with different number of gpus:

![1 gpu](https://user-images.githubusercontent.com/69603395/165116072-b23b2ad4-f432-4eed-a7cb-e535e54c5c84.png)
![2 gpus](https://user-images.githubusercontent.com/69603395/165116079-2b86fc32-da1a-4116-8efe-614a07002755.png)
![4 gpus](https://user-images.githubusercontent.com/69603395/165116084-92d7657b-16d4-4aa2-ba81-ae7ebdccbce8.png)
![6 gpus](https://user-images.githubusercontent.com/69603395/165116086-14effd17-2b7e-446c-9327-16bb759e5db3.png)

I think it is curious when we raise the number of gpus, each gpu work `less` but as a group of gpus obviously is better, since is the sum of every gpu's work.

**Comment by Setol21** [comment]
### 1 GPU:
Img/sec : 260.7

### 2 GPUs:
Each gpu img/sec : 245.4
Every gpu img/sec : 490.8

### 4 GPUs:
Each gpu img/sec : 232
Every gpy img/sec : 928.2

### 6 GPUs:
Each gpu img/sec : 227
Every gpy img/sec : 1365.3

**Comment by pescap** [comment]
> For the `tensorflow2_keras_mnist.py` file:
> 
> ## General
> 1. I managed to understand that this code runs in every gpu, therefore when `hvd.rank()` is used it gives the number of the worker (gpu). So when `hvd.rank()==0` appears on the code it means that part of the code belongs only for the worker 0.

That's it! 
`horovod` is very similar to traditional `MPI` code structure.  

> ## Callbacks:
> 1. In the part of callbacks, `hvd.callbacks.BroadcastGlobalVariablesCallback(0)` makes sure the initial variables are broadcasted to the rest of workers. It is super important when working with random or checkpoint (previous trained/saved) data. The other command `hvd.callbacks.MetricAverageCallback()` averages the metrics from all workers (to work like a synchronized team).
Nice! If we do not need checkpoints, is it important to use these callbacks? 

> 2. `hvd.callbacks.LearningRateWarmupCallback(initial_lr=scaled_lr, warmup_epochs=3, verbose=1)` This part was an _exclusive_ for this code, when training it, it showed worse accuracy when not scaling the learning rate in the first 3 epochs, but I think this is something to come with when searching for a good accuracy.

**Comment by pescap** [comment]
> @pescap
> 
> Benchmarking tensorflow2_keras_synthetic_benchmark.py with different number of gpus:
> 
> ![1 gpu](https://user-images.githubusercontent.com/69603395/165116072-b23b2ad4-f432-4eed-a7cb-e535e54c5c84.png) ![2 gpus](https://user-images.githubusercontent.com/69603395/165116079-2b86fc32-da1a-4116-8efe-614a07002755.png) ![4 gpus](https://user-images.githubusercontent.com/69603395/165116084-92d7657b-16d4-4aa2-ba81-ae7ebdccbce8.png) ![6 gpus](https://user-images.githubusercontent.com/69603395/165116086-14effd17-2b7e-446c-9327-16bb759e5db3.png)
> 
> I think it is curious when we raise the number of gpus, each gpu work `less` but as a group of gpus obviously is better, since is the sum of every gpu's work.

Check: https://github.com/ComputationalRadiationPhysics/NeuralSolvers

**Comment by pescap** [comment]
@agustinesoto @Setol21 

Check the code I added so far. 

From one side: 
https://github.com/pescap/FastDeepXDE/blob/main/examples/horovod/tensorflow2_mnist_light1.py

is a very simple DL case, with tensorflow 2.0, and mnist data. Read the comments to understand how I arrived at next versions. I have been trying to implement horovod on this simple example -> mnist_light2, and then mnist_light2.

We need to understand exactly how to replace `model.fit` by the horovod function for training, and obtain the same results with horovod (and 1 GPU) v/s without horovod (mnist_light.py). 

From the other side:
https://github.com/pescap/FastDeepXDE/blob/main/examples/horovod/horovod_tensorflow2_mnist.py

is the classic horovod tutorial. It would be interesting to create a version of this code without horovod (and compare horovod 1GPU v/s without horovod).

Does this make sense to you?

**Comment by Setol21** [comment]
> 

I'm interpreting that we need to understand how to implement the fitting part of the code to match the results with horovod and 1 GPU and without horovod (original sccript). And the reverse, remove horovod and compare it.

Am I right?

**Comment by pescap** [comment]
More references:

https://stackoverflow.com/questions/55008213/tensorflow-mirror-strategy-and-horovod-distribution-strategy

https://github.com/leondgarse/Keras_insightface/discussions/17
